### PR TITLE
Add an optional approximation parameter to skip splits in tree search

### PR DESCRIPTION
**Pull request by erikcs**
Add `split_step`, an optional approximation parameter that controls how many splits to consider.

`split_step` = 1 considers splitting at each distinct value (default).

`split_step` = 10 considers splitting at every 10'th distinct value.

(did not add the approximation to `level_zero_learning` as that seemed unnecessary, and it lets edge case input pass throw nicely)

Example:

```R
set.seed(1)
p = 5; n = 40000; prop = 0.2
X = matrix(runif(n*p),n,p)
W = rbinom(n, 1, prop)
tau = 50 * (2 * X[,2] - (1 - X[,1])^4 - 0.5)
mu0 = 50 * (0.5 * (X[,3] - X[,4]))
Y = mu0 + W * tau + runif(n, -80, 80)
IPW.score = cbind((1 - W) / (1 - prop) * Y, W / prop * Y)

# rounding down to ca 100 distinct values
system.time(opt.tree <- policy_tree(round(X, 2), IPW.score, depth = 2))
#  user  system elapsed 
# 16.365   0.079  16.482 
opt.reward <- mean(IPW.score[cbind(1:n, predict(opt.tree, round(X, 2)))])
opt.reward
#[1] 21.90584

# only consider every 40 000 / 400 = 100'th point when splitting
system.time(skip.tree <- policy_tree(X, IPW.score, depth = 2, split.step = 400))
#  user  system elapsed 
# 16.096   0.081  16.216 
skip.reward <- mean(IPW.score[cbind(1:n, predict(skip.tree, X))])
skip.reward
#[1] 21.91867
```

### CODE REVIEW

**Comment by erikcs** [comment]
@halflearned split.step = 2 may be of interest in speeding up bandit policy learning

File: R/policy_tree.R [hunk 1]
```diff
@@ -7,6 +7,10 @@
 #' @param X The covariates used. Dimension \eqn{Np} where \eqn{p} is the number of features.
 #' @param Gamma The rewards for each action. Dimension \eqn{Nd} where \eqn{d} is the number of actions.
 #' @param depth The depth of the fitted tree. Default is 2.
+#' @param split.step An optional approximation parameter, the number of possible splits to consider when
+#'  performing tree search. split.step = 1 (default) considers every possible split, split.step = 10
+#'  considers splitting at every 10'th point and may yield a substantial speedup for densely packed
```
Comment (kanodiaayush) [inline R/policy_tree.R]: s/may/will

File: R/policy_tree.R [hunk 2]
```diff
@@ -27,7 +31,7 @@
 #' @useDynLib policytree
 #' @importFrom Rcpp evalCpp
 #' @importFrom utils type.convert
-policy_tree <- function(X, Gamma, depth = 2) {
+policy_tree <- function(X, Gamma, depth = 2, split.step = 1) {
```
Comment (kanodiaayush) [inline R/policy_tree.R:34]: why split.step and not split_step?

File: src/tree_search.cpp [hunk 3]
```diff
@@ -306,8 +318,13 @@ std::unique_ptr<Node> find_best_split(const std::vector<flat_set>& sorted_sets,
         if (point_bk.get_value(p) >= next->get_value(p)) { // are the values the same then skip
           continue;
         }
-        auto left_child = find_best_split(left_sorted_sets, level - 1, data, sum_array);
-        auto right_child = find_best_split(right_sorted_sets, level - 1, data, sum_array);
+        if (++split_counter == split_step) { // only split at every `split_step`th point
+          split_counter = 0;
+        } else {
+          continue;
```
Comment (kanodiaayush) [inline src/tree_search.cpp:326]: split_counter+=1 here

File: src/tree_search.cpp [hunk 4]
```diff
@@ -306,8 +318,13 @@ std::unique_ptr<Node> find_best_split(const std::vector<flat_set>& sorted_sets,
         if (point_bk.get_value(p) >= next->get_value(p)) { // are the values the same then skip
           continue;
         }
-        auto left_child = find_best_split(left_sorted_sets, level - 1, data, sum_array);
-        auto right_child = find_best_split(right_sorted_sets, level - 1, data, sum_array);
+        if (++split_counter == split_step) { // only split at every `split_step`th point
```
Comment (kanodiaayush) [inline src/tree_search.cpp]: s/++split_counter/split_counter

**Review by kanodiaayush — CHANGES_REQUESTED** [review]

**Review by kanodiaayush — APPROVED** [review]

**Pull request merged**